### PR TITLE
complete feat/stack_data_structure

### LINKED ISSUE #58: Implement Stack data structure
# Implement Stack data structure in Cairo

## Objective
Create implementation of Stack (LIFO) data structure in Cairo.

For a better guidance if needed you can check this: https://github.com/keep-starknet-strange/alexandria/blob/main/packages/data_structures/src/stack.cairo

## Stack Features
- push(element)
- pop()
- peek()
- is_empty()
- size()

## Technical Considerations
- Overflow/underflow control
- Complete unit tests

## 📜 Additional Notes
⚠ **Do not apply until the ODHack begins.**
⚠ **Read our guidelines before applying.**

### PR DESCRIPTION
**Pull request by KevinMB0220**
Closes #58 

# 🛠️ Implementation of `MemoryVec` in Cairo

## Description
A `MemoryVec` structure has been implemented in Cairo to efficiently manage in-memory vectors. This structure uses a `Felt252Dict` with `Nullable` elements to store data and provides methods that replicate typical vector functionalities, such as adding, accessing, and removing elements.

## Key Features
- **Initialization**: `new` method to create an empty vector.
- **Access**:
  - `get`: Returns an optional element by index.
  - `at`: Returns an element directly; throws an error if the index is out of bounds.
- **Modification**:
  - `push`: Adds an element to the end of the vector.
  - `set`: Updates the value at a specific position.
  - `pop`: Removes and returns the last element.
  - `peek`: Returns the last element without removing it.
- **Properties**:
  - `len`: Returns the current length of the vector.
  - `is_empty`: Checks if the vector is empty.
- **Destruction**: `Destruct` implementation ensures proper resource release by "squashing" the data.

## Use of Traits
The implementation uses the following traits:
- `MemoryVecTrait`: Defines the main operations for the vector.
- `NullableTrait`: Handles null values within the dictionary.
- `WrappingAdd`: Ensures safe length increments without overflow.
## Tests pased
image

### CODE REVIEW

**Review by coxmars — COMMENTED** [review]
@KevinMB0220 add a .tools-versions file in the scarb project "cairo/scripts/stack_data_structure" to document your scarb version or other dependencies.

**Comment by coxmars** [comment]
@KevinMB0220 can you add a small example using this data structure? It's to validate the functionality and provide a brief context for new devs who want to use it.

**Comment by KevinMB0220** [comment]
> @KevinMB0220 can you add a small example using this data structure? It's to validate the functionality and provide a brief context for new devs who want to use it.

Ready

**Comment by KevinMB0220** [comment]
> @KevinMB0220 can you add a small example using this data structure? It's to validate the functionality and provide a brief context for new devs who want to use it.

ready

File: cairo/scripts/stack_data_structure/.tools-versions
Comment (danielcdz) [inline cairo/scripts/stack_data_structure/.tools-versions:1]: this file needs to be named:`.tool-versions` to works as expected

**Review by danielcdz — CHANGES_REQUESTED** [review]

**Review by danielcdz — COMMENTED** [review]
Some files need an extra line at the end, can you please add it?

**Comment by KevinMB0220** [comment]
> Some files need an extra line at the end, can you please add it?

Ready

File: cairo/scripts/stack_data_structure/.tools-versions
Comment (KevinMB0220) [reply cairo/scripts/stack_data_structure/.tools-versions:1]: Ready

**Review by danielcdz — APPROVED** [review]

**Pull request merged**